json-fleece-openapi3/src/Fleece/OpenApi3/Schemas/Schemas.hs

```diff
@@ -0,0 +1,209 @@
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+module Fleece.OpenApi3.Schemas.Schemas
+  ( FleeceOpenApi3
+  , nonEmpty
+  , set
+  , nonEmptyText
+  , integer
+  , unboundedIntegralNumberNamed
+  , unboundedIntegralNumber
+  , boundedIntegralNumberNamed
+  , boundedIntegralNumber
+  , int
+  , int8
+  , int16
+  , int32
+  , int64
+  , word
+  , word8
+  , word16
+  , word32
+  , word64
+  , double
+  , float
+  , utcTime
+  , utcTimeWithFormat
+  , localTime
+  , localTimeWithFormat
+  , zonedTime
+  , zonedTimeWithFormat
+  , day
+  , dayWithFormat
+  , dateTimeFormat
+  , customFormat
+  ) where
+
+import qualified Data.Int as I
+import qualified Data.List.NonEmpty as NEL
+import Data.Maybe (fromJust)
+import qualified Data.NonEmptyText as NET
+import qualified Data.Set as Set
+import qualified Data.Text as T
+import qualified Data.Time as Time
+import Data.Typeable (Typeable)
+import qualified Data.Vector as V
+import qualified Data.Word as W
+import qualified Fleece.Core as FC
+
+import Fleece.OpenApi3.Schemas.OpenApi3Validator (OpenApi3Validator (maximumIntegral, minItems, minLength, minimumIntegral, setValidatorFormat, setValidatorType, uniqueItems))
+
+type FleeceOpenApi3 schema =
+  ( FC.Fleece schema
+  , OpenApi3Validator (FC.Validator schema)
+  )
```

Comment:
A consequence of implementing things this way is that users have to replace their `Fleece` constraints with `FleeceOpenApi3` constraints if they want to use the validator. An alternative would be to move the `OpenApi3Validator` class to `json-fleece-core` and export a constraint like this that's just called `Fleece`, potentially renaming the existing class. In other words, we'd be saying that OpenAPI 3 validations are the default. I didn't do that initially because it seemed too opinionated, but maybe it's a better approach.

Reply:
If it's _only_ if they want to use the validator that seems fine. I don't think we'd want to demand OpenApi3 (or any) participation for _all_ of Fleece, though, if that's what you mean by the alternative.

Tangentially it appears that we only have one API client that isn't using OpenAPI 3 -- tai software (they have two versions of their API and for both we are using swagger codegen). I messages them to see if they have an OpenAPI 3 spec.

Reply:
The constraint `OpenApi3Validator (FC.Validator schema)` means that the validators you write can use the methods on the `OpenApi3Validator` class, which are only related to OpenAPI 3 in that they represent the validators supported by the OpenAPI 3 spec, i.e. you don't have to be using OpenAPI 3 anywhere else to use the validators.

So in that sense, the alternative of placing the `OpenApi3Validator` class and the constraint above in `json-fleece-core` doesn't demand OpenAPI 3 participation, it just enables you to write validators in terms of those supported by OpenAPI 3. 

Given @ysangkok's point above about the transitive `openapi3` dependency I introduced into all of the `fleece` packages, it would actually be less demanding of participation to move these modules into `json-fleece-core`.

The idea would be to:
1. rename the `Fleece` class to something like `FleeceSchema`
2. export a constraint like the above called `Fleece`

We could potentially rename `OpenApi3Validator` to something else as well, though I'm not sure that there another name that makes more sense. Maybe the methods should just be part of the `FleeceValidator` class instead.

Reply:
Actually there are good reasons to avoid making them part of `FleeceValidator`, e.g. `type` and `format` values as `Text` don't necessarily have interpretations outside of the context of OpenAPI 3.